motionplan/tpSpaceRRT.go

```diff
@@ -0,0 +1,435 @@
+	sinceLastNode := 0.
+	_ = sinceLastNode
+	lastDist := 0.
+
+	// Check each point along the trajectory to confirm constraints are met
+	for _, trajPt := range trajK {
+		if trajPt.Dist > randD {
+			// After we've passed randD, no need to keep checking, just add to RRT tree
+			break
+		}
+		sinceLastCollideCheck += (trajPt.Dist - lastDist)
+		trajState := &State{Position: spatialmath.Compose(cNode.endConfig, trajPt.Pose), Frame: mp.frame}
+		if sinceLastCollideCheck > mp.planOpts.Resolution {
+			ok, _ := mp.planOpts.CheckStateConstraints(trajState)
+			if !ok {
+				pass = false
+				break
+			}
+		}
+
+		lastPose = trajState.Position
+		lastNode = trajPt
+		lastDist = lastNode.Dist
+	}
+	if pass {
+		// add the last node in trajectory
+		successNode = &configurationNode{
+			endConfig: lastPose,
+			q:         referenceframe.FloatsToInputs([]float64{float64(ptgNum), float64(randK), lastNode.Dist}),
+			cost:      cNode.Cost() + lastNode.Dist,
+		}
+	} else {
+		return nil
+	}
+	if tryGoal && bestDist < mp.planOpts.GoalThreshold {
+		// If we tried the goal and have a close-enough XY location, check if the node is good enough to be a final goal
+		acceptableGoal = mp.planOpts.GoalThreshold > mp.planOpts.goalMetric(&State{Position: lastPose, Frame: mp.frame})
+	}
+
+	cand := &candidate{dist: bestDist, treeNode: cNode, newNode: successNode, finalGoal: acceptableGoal}
+
+	if !acceptableGoal && successNode != nil {
+		// check if this  successNode is too close to nodes already in the tree, and if so, do not add.
+
+		// Get nearest neighbor to new node that's already in the tree
+		utils.PanicCapturingGo(func() {
+			nm.nearestNeighbor(ctx, mp.planOpts, successNode, rrt.maps.startMap, nnchan)
+		})
+		nearest := <-nnchan
+		dist := math.Inf(1)
+		if nearest != nil {
+			nearNode, _ := nearest.(*configurationNode)
+			dist = mp.planOpts.DistanceFunc(&Segment{StartPosition: successNode.endConfig, EndPosition: nearNode.endConfig})
+		}
+
+		// Ensure successNode is sufficiently far from the nearest node already existing in the tree
+		// If too close, don't add a new node
+		if dist < 5. {
+			cand = nil
+		}
+	}
+	return cand
+}
+
+func (mp *tpspaceRRTMotionPlanner) extendMap(
+	ctx context.Context,
+	candidateNodes map[float64][2]*configurationNode,
+	rrt *rrtParallelPlannerShared,
+	tpFrame tpspace.PTGProvider,
+) *configurationNode {
+	var addedNode *configurationNode
+	// If we found any valid nodes that we can extend to, find the very best one and add that to the tree
+	if len(candidateNodes) > 0 {
+		bestDist := math.Inf(1)
+		var bestNode [2]*configurationNode
+		for k, v := range candidateNodes {
+			if k < bestDist {
+				bestNode = v
+				bestDist = k
+			}
+		}
+
+		ptgNum := int(bestNode[1].Q()[0].Value)
+		randK := uint(bestNode[1].Q()[1].Value)
+
+		trajK := tpFrame.PTGs()[ptgNum].Trajectory(randK)
+
+		lastDist := 0.
+		sinceLastNode := 0.
+
+		for _, trajPt := range trajK {
+			select {
+			case <-ctx.Done():
```

Comment:
[q] this is checking for a cancelled context?
It looks strange having a default case which does nothing.
can we instead do: `if ctx.Err() != nil`